### ISSUE TITLE
Model estimation on a subset of subjects and sessions

### ISSUE THREAD

**Issue by TamerGezici** [automaticanalysis/automaticanalysis#381]
I have preprocessed data, which have a list of subjects and sessions in the script:

```
% prep script
subj_list = {'sub-01','sub-02','sub-03','sub-04','sub-05','sub-06','sub-07','sub-08','sub-09','sub-11','sub-12','sub-13','sub-14','sub-19','sub-20','sub-21','sub-22','sub-23','sub-24','sub-25','sub-26','sub-27','sub-28','sub-29','sub-30','sub-31','sub-32','sub-33','sub-34','sub-35','sub-36','sub-37','sub-38','sub-39','sub-40','sub-41','sub-42','sub-43','sub-44','sub-45','sub-46','sub-47','sub-48','sub-49','sub-50','sub-51','sub-52','sub-53','sub-54','sub-55'};`
tasknames = {'discrA_run-01','discrA_run-02','discrA_run-03','discrB_run-01','discrB_run-02','discrB_run-03'};
for i = 1:size(tasknames,2)
    aap = aas_addsession(aap,tasknames{i}); 
end
```

Now I am trying to do a pipeline connection to do GLM (only GLM! No contrasts), on a specific subset of subjects and subset of sessions:

```
% GLM script
subj_list = {'sub-03','sub-04','sub-06','sub-11','sub-20','sub-21','sub-22','sub-23','sub-24','sub-25','sub-26','sub-27','sub-28','sub-29','sub-30','sub-33','sub-35','sub-36','sub-38','sub-39','sub-40','sub-47','sub-48','sub-49','sub-50','sub-51','sub-52','sub-53','sub-54','sub-55'};
tasknames = {'discrA_run-01','discrA_run-02','discrA_run-03'};
for i = 1:size(tasknames,2)
    aap = aas_addsession(aap,tasknames{i}); 
end
```

However, it skips some subjects:

```
MODULE aamod_firstlevel_model RUNNING: First level model for D:\studies\analysis\tamer_study3\Ipek\GLM_results\mvpa_button_press_A\aamod_firstlevel_model_00001\sub-03
WARNING: No inputs selected for stream epi. --> MODULE aamod_firstlevel_model will be SKIPPED

	 	 	 Elapsed time = 0h:0m:0s
	 	 	 ...at = 08:35:16 (26/3/2023)
MODULE aamod_firstlevel_model RUNNING: First level model for D:\studies\analysis\tamer_study3\Ipek\GLM_results\mvpa_button_press_A\aamod_firstlevel_model_00001\sub-04
WARNING: No inputs selected for stream epi. --> MODULE aamod_firstlevel_model will be SKIPPED
```

It doesn't skip sub-06. 

You will notice that in the longer subj_list (prep) sub-03 is the third subject, however in the GLM list, it is the first. 

nstead of iterating over the new subj_list, it iterates over the old one. So it goes like:

this is subj-01, and can not match the session name (discrA) so it skips it. Same story for sub-02. However when it comes to sub-03, it thinks that it is sub-06, which also does have that session (it is the third subject in the larger list) it starts processing it. But due to this skipping, subj-03 and sub-04 can not be processed. Since sub-06 is the third subject in the GLM script, and the third subject is a match in the prep script so it doesn't skip sub-06.

So basically, for some reason is stuck with the old subject list, so it tries matching indices based on the larger subject list. 

I believe the issue may be related to some code here:

https://github.com/automaticanalysis/automaticanalysis/blob/f7849b18297c277e3603e06e35010a437520d2eb/aa_engine/aa_doprocessing_onetask.m#L156-L171

**Comment by TamerGezici** [comment]
Just now It did occur to me that since all subjects belong to a specific block, I could do something like this:

Adding all subjects, but specifiying specific sessions. 
```
subj_list = {'sub-01','sub-02','sub-03','sub-04','sub-05','sub-06','sub-07','sub-08','sub-09','sub-11','sub-12','sub-13','sub-14','sub-19','sub-20','sub-21','sub-22','sub-23','sub-24','sub-25','sub-26','sub-27','sub-28','sub-29','sub-30','sub-31','sub-32','sub-33','sub-34','sub-35','sub-36','sub-37','sub-38','sub-39','sub-40','sub-41','sub-42','sub-43','sub-44','sub-45','sub-46','sub-47','sub-48','sub-49','sub-50','sub-51','sub-52','sub-53','sub-54','sub-55'};
tasknames = {'discrA_run-01','discrA_run-02','discrA_run-03'};
```

**Comment by tiborauer** [comment]
Have you used debugging to see which remote data is looked for which local subject?
I.e., What is `aas_doneflag_getpath_bydomain(sourceaap,inp.sourcedomain,deps{d}{2},sourcenumber)`? Does it match with the local subject?

**Comment by TamerGezici** [comment]
In my opinion, whatever is causing this issue is mainly because somewhere in aap the subject list is the remote subject list. So subjects who do not exist in the local list should be removed, and a new subject list must be formed.

**Comment by tiborauer** [comment]
Yes! Indeed, the remote subject list, which is stored in the (copied) remote aap, causes the issue. aa usually uses subject index rather than subject name as a reference, and the indices are obviously not the same. It is also true for the sessions; however, in your case, the indices are the same because you use the first 3 sessions, whose indices are the same in the remote aap.

**Comment by TamerGezici** [comment]
To me, it looks fine.

It returns a struct with the field:

K>> map.subject

ans =

     3     4     6    10    15    16    17    18    19    20    21    22    23    24    25    28    30    31    33    34    35    42    43    44    45    46    47    48    49    50

And some other fields as well:

![image](https://user-images.githubusercontent.com/11014942/229178023-c16f842e-9e26-4a73-9d21-1fa2130d17ad.png)


Where do I place this in `aa_doprocessing_onetask`? Not sure how to use these indexes.

It also returns a session index, the thing is I never had an issue with sessions but I will check that as well. (removing A sessions and trying to process only B sessions)

**Comment by tiborauer** [comment]
In lines 158 and 161, `deps{d}{2}` corresponds to the indices. So, before 155, you may want to update it with the appropriate indices.

**Comment by TamerGezici** [comment]
When forming a pipeline connection, Is the remote aap present anywhere in the current aap? Shall I write code so that the remote aap is loaded (similar to beforeuserchanges) to the aap structure if we are forming a pipeline connection?

**Comment by tiborauer** [comment]
It is `sourceaap` in https://github.com/automaticanalysis/automaticanalysis/blob/f7849b18297c277e3603e06e35010a437520d2eb/aa_engine/aa_doprocessing_onetask.m#L144

**Comment by TamerGezici** [comment]
Hi Tibor, sorry for the late reply! I got caught up with some other things and left this hanging. 

Back then when I tried your implementation, I saw that it was a little too complicated to implement so I had given up. Now again I stumbled upon the same issue and tried fixing it...

I can implement a fix for mapping subjects, which I think is very straightforward.

However, it looks like even if you update parts that are related to sessions in `deps` before these lines for the "remote" indices:

Assume the remote indices are 4,5,6 but in the local script they are 1,2,3. Here, you can sort them by mapping the indices.

https://github.com/automaticanalysis/automaticanalysis/blob/f7849b18297c277e3603e06e35010a437520d2eb/aa_engine/aa_doprocessing_onetask.m#L156-L165

Then they need to reverted to the "local" indices here: (1,2,3) because that is what is expected.

https://github.com/automaticanalysis/automaticanalysis/blob/e1796d20a222ebdfc843b4ae79b8b299f4738ff8/aa_engine/aas_retrieve_inputs_part1.m#L31

Because they are needed here:

https://github.com/automaticanalysis/automaticanalysis/blob/e1796d20a222ebdfc843b4ae79b8b299f4738ff8/aa_engine/aas_getpath_bydomain.m#L22

And also here:

https://github.com/automaticanalysis/automaticanalysis/blob/e1796d20a222ebdfc843b4ae79b8b299f4738ff8/aa_engine/aas_retrieve_inputs_part1.m#L68

The issue is, in later iterations of "aa_doprocessing_onetask", the deps{d} contains different domains, so checks must be added so that we don't accidentally index the array out of bounds.

I know it sounds complicated, and it is. I spent some good hours trying to figure out a way, but I can not see anyother way to fix this without refactoring all these functions somehow.

**Comment by tiborauer** [comment]
I am unsure I understand the issue:
As far as I understand you need two sets of remote indices: original and mapped-to-local. You need to use the original indices with the remote app and the mapped-to-local indices with the local aap.

**Comment by TamerGezici** [comment]
I came up with a better solution. Will make a PR soon.
